ledger/ledger.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package ledger implements a modified map with three unique characteristics:
+// 1. every unique state of the map is given a unique hash
+// 2. prior states of the map are retained for a fixed period of time
+// 2. given a previous hash, we can retrieve a previous state from the map, if it is still retained.
+package ledger
+
+import (
+	"encoding/base64"
+	"time"
+)
+
+// Ledger exposes a modified map with three unique characteristics:
+// 1. every unique state of the map is given a unique hash
+// 2. prior states of the map are retained for a fixed period of time
+// 2. given a previous hash, we can retrieve a previous state from the map, if it is still retained.
+type Ledger interface {
+	// Put adds or overwrites a key in the Ledger
+	Put(key, value string) (string, error)
+	// Delete removes a key from the Ledger, which may still be read using GetPreviousValue
+	Delete(key string) error
+	// Get returns a the value of the key from the Ledger's current state
+	Get(key string) (string, error)
+	// RootHash is the hash of all keys and values currently in the Ledger
+	RootHash() string
+	// GetPreviousValue executes a get against a previous version of the ledger, using that version's root hash.
+	GetPreviousValue(previousRootHash, key string) (result string, err error)
+}
+
+type smtLedger struct {
+	tree smt
```

Comment:
One last comment, I think you should change this composition. Specifically, this field should be a pointer (so `tree *smt), the Make function should return a smtLedger and not the address of same, and all the functions in this file should accept an smtLedger instead of a pointer to same.

The reason that right now, this code is doing a copy of the allocated smt object into the smtLedger. It's bad practice, and sometimes fatal, to copy a struct that includes a mutex (such as smtLedger).  So instead, you allocate the smt object in the heap and keep a pointer to it, and return a wrapper struct over that pointer.

Reply:
As I understand it, removing the pointer from the receiver functions would mean that each function gets a copy of the smtLedger, which means the core object is never modified.  Tests bear this out.  I don't think that's what we're aiming for, is it?

Reply:
That's not the construct I'm suggesting. What I mean is:

```go
type smtLedger struct {
    tree *smt
}

func (l smtLedger) Put() 
```

The core smt data structure is allocated once and never copied. The pointer to it is the only thing that gets copied. This is totally safe and reasonable.

Reply:
Ok, got it.